Dockerfile

```diff
@@ -1,31 +1,47 @@
-FROM alpine:3.9
+################################################################################
+# Builder                                                                      #
+################################################################################
+FROM alpine:3.9 as builder
 
 MAINTAINER Alexey Kutepov <reximkut@gmail.com>
 
-VOLUME ["/tmp/hypernerd"]
-
-RUN apk add --no-cache openssl-dev
-RUN apk add --no-cache curl
-RUN apk add --no-cache ghc
-RUN apk add --no-cache cabal
-RUN apk add --no-cache gmp-dev
-RUN apk add --no-cache build-base
-RUN apk add --no-cache zlib-dev
-
-RUN cabal update
-RUN mkdir /hypernerd/
-COPY ./HyperNerd.cabal /hypernerd/
-COPY ./Setup.hs /hypernerd/
-WORKDIR /hypernerd/
-RUN cabal sandbox init
+# Have base OS package installation in a single layer
+RUN \
+  apk update && \
+  apk add --no-cache \
+    openssl-dev curl ghc cabal \
+    gmp-dev build-base zlib-dev && \
+  cabal update && \
+  mkdir -p /build/
+
+WORKDIR /build/
+COPY . /build/
+
+# Install necessary dependencies and build HyperNerd
 # NOTE: happy package provides an executable that is needed to build
 # qm-interpolated-string package. qm-interpolated-string does not
 # depend on happy for some reason and that's why we have to install it
 # separately.
-RUN cabal install happy-1.19.9
-RUN cabal install --only-dependencies
-COPY ./resources/ /hypernerd/resources/
-COPY ./src/ /hypernerd/src/
-RUN cabal build exe:HyperNerd
+RUN \
+  cabal install happy-1.19.9 && \
+  cabal install -j4 --only-dependencies && \
```

Comment:
1. Why do you think I have enough cores on my deployment machine to support `-j4`?
2. `cabal install --only-dependencies` is the longest operation (up to 30 minutes on my machine) and the entire Dockerfile was specifically designed to reduce its impact.

Originally we `COPY`-ed only `HyperNerd.cabal` without copying the whole codebase (this is very important!) before installing the dependencies. And if the `HyperNerd.cabal` was not changed (which happens rarely) the docker won't rebuild the step with installing the dependencies which on average made the whole build faster.

Now this Dockerfile will ALWAYS (as fas as I can understand) install the dependencies no matter what. How is that an improvement?

Reply:
Relevant issue #510